doc/manual/variables-and-scoping.rst

```diff
@@ -6,234 +6,386 @@
+*local* variables::
+
+    x,y = 1,2
+    function foo()
+        x = 2 # introduces a new local
+        function bar()
+            x = 10 # modifies the parent's x
+            return x+y # y is global
+        end
+        return bar() + x # 12 + 10 (x is modified in call of bar())
+    end
+
+    julia> foo()
+    22  # (x,y unchanged)
+
+The distinction between inheriting global and local variables for
+assignment can lead to some slight differences between functions
+defined in local vs. global scopes.  Consider the modification of the
+last example by moving ``bar`` to the global scope::
+
+    x,y = 1,2
+    function bar()
+        x = 10 # local
+        return x+y
+    end
+    function foo()
+        x = 2 # local
+        return bar() + x # 12 + 2 (x is not modified)
+    end
+
+    julia> foo()
+    14 # as x is not modified anymore.
+       # (x,y unchanged)
+
+Note that above subtlety does not pertain to type and macro
+definitions as they can only appear at the global scope.
+There are special scoping rules concerning the evaluation of default
+and keyword function arguments which are described in the
+:ref:`Function section <man-evaluation-scope-default-values>`.
+
+
+An assignment introducing a variable used inside a function, type or
+macro definition need not come before its inner usage:
+
+.. doctest::
+
+    julia> f = y -> x + y
+    (anonymous function)
+
+    julia> f(3)
+    ERROR: UndefVarError: x not defined
+     in anonymous at none:1
+
+    julia> x = 1
+    1
+
+    julia> f(3)
+    4
+
+This behavior may seem slightly odd for a normal variable, but allows
+for named functions — which are just normal variables holding function
+objects — to be used before they are defined. This allows functions to
+be defined in whatever order is intuitive and convenient, rather than
+forcing bottom up ordering or requiring forward declarations, as long
+as they are defined by the time they are actually called.  As an
+example, here is an inefficient, mutually recursive way to test if
+positive integers are even or odd::
+
+    even(n) = n == 0 ? true  :  odd(n-1)
+    odd(n)  = n == 0 ? false : even(n-1)
+
+    julia> even(3)
+    false
+
+    julia> odd(3)
+    true
+
+Julia provides built-in, efficient functions to test for oddness and evenness
+called :func:`iseven` and :func:`isodd` so the above definitions should only be
+taken as examples.
+
+Hard vs. Soft Local Scope
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Blocks which introduce a soft local scope, such as loops, are
+generally used to manipulate the variables in their parent scope.
+Thus their default is to fully access all variables in their parent
+scope.
+
+Conversely, the code inside blocks which introduce a hard local scope
+(function, type and macro definitions) can be executed at any place in
+a program.  Remotely changing the state of global variables in other
+modules should be done with care and thus this is an opt-in feature
```

Comment:
OK thank you @mauro3, I misinterpreted "requiring the global keyword": to modify `A.a` from module `B`, using `global` within `B` won't help: only module `A` can provide the means, via `global` and a function/macro, to modify `A.a`. I think this sentence is a bit ambiguous and appeared contradictory to what was explained above: "Remotely changing the state of global variables in other modules [...] requir[es] the ``global`` keyword". So I would suggest you give this example (or I missed a similar one?), as otherwise I feel stupid not to understand by myself how to modify a global in `A` from `B` using the `global` keyword.